### PR TITLE
Remove repositories

### LINKED ISSUE #672: Internal error (500) when creating incident via API request (curl)
- why does the error occur in my case? (i'm using the latest docker container)
- an API error response should also be using JSON content-type

```bash
$ curl -H 'Accept: */*' -H "accept-encoding: gzip" -H "content-type: application/json" -H 'X-Cachet-Token: xyz' -X POST http://xxx.yyyy.zzz/api/v1/incidents -d '
{
  name:"Test name",
  message:"Test message",
  status: 1
}
'
<!doctype html>
<html lang="en-GB">
<head>
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge,chrome=1">
<title>Error 500</title>
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<style type="text/css">
    @import  url(//fonts.googleapis.com/css?family=Droid+Sans);
    html{font-size:100%;-webkit-text-size-adjust:100%;-ms-text-size-adjust:100%;font-family:sans-serif;color:#222;}
    body{font-family:'Droid Sans', sans-serif;font-size:10pt;color:#555;line-height:25px;margin:0;}
    a{color:#00e;}
    a:visited{color:#551a8b;}
    a:hover{color:#72ADD4;}
    a:focus{outline:thin dotted;}
    a:hover,a:active{outline:0;}
    hr{display:block;height:1px;border:0;border-top:1px solid #ccc;margin:1em 0;padding:0;}
    .wrapper{width:760px;margin:0 auto 5em;}
    .error-spacer{height:4em;}
    a,a:visited{color:#2972A3;}
    .row-fluid [class*="span"]{float:left;width:100%;margin-left:2.0744680851064%;-webkit-box-sizing:border-box;-moz-box-sizing:border-box;box-sizing:border-box;}
    .row-fluid [class*="span"]:first-child{margin-left:0;}
    .row-fluid .span3{width:99.946808510638%;}
    .row-fluid .span2{width:57.393617021277%;}
    .row-fluid .span1{width:40.372340425532%;}
    .pull-right{float:right;}
    .ellipsis{overflow: hidden;white-space: nowrap;text-overflow: ellipsis;-o-text-overflow: ellipsis;}
</style>
<link rel="shortcut icon" href="http://xxx.yyyy.zzz/favicon.ico">
</head>
<body>
<div class="wrapper">
    <div class="error-spacer"></div>
    <div role="main" class="main">
        <div class="row-fluid">
            <div class="span3">
                <div class="span1 ellipsis">
                    <h1>Error 500</h1>
                </div>
                <div class="span2 ellipsis">
                    <h2 class="pull-right">Houston, We Have A Problem</h2>
                </div>
            </div>
            <h2>Internal Server Error</h2>
            <hr>

            <h3>What does this mean?</h3>

            <p>
                Something went wrong on our servers while we were processing your request.
                An error has occurred and this resource cannot be displayed.
                This occurrence has been logged, and a highly trained team of monkeys has been
                dispatched to deal with your problem. We're really sorry about this, and will
                work hard to get this resolved as soon as possible.
            </p>
            <p>
                Perhaps you would like to go to our <a href="http://xxx.yyyy.zzz/">home page</a>?
            </p>
        </div>
    </div>
</div>
</body>
</html>
```

### PR DESCRIPTION
**Pull request by jbrooksuk**
All of the creation code needs to be swapped out for the usual model code.

Closes #672 #692 #658

### CODE REVIEW

**Comment by jbrooksuk** [comment]
Also we need the `AbstractApiController` to handle returning of validation too. So if you send through a bad dataset and the model isn't created properly, then it should return an error.

**Comment by jbrooksuk** [comment]
Rebased.

**Comment by jbrooksuk** [comment]
Needs more thorough testing too.

**Comment by Lavoaster** [comment]
Out of curiosity, why are you abandoning the repository pattern?

**Comment by GrahamCampbell** [comment]
Because it doesn't provide us any real value.

**Comment by jbrooksuk** [comment]
@Lavoaster exactly what Graham said. We will actually be re-introducing this in V2 since it'll make a lot more sense then.

**Comment by jbrooksuk** [comment]
All that is left now is to thoroughly test it.

**Comment by jbrooksuk** [comment]
Maybe we should also add some tests for the API itself?

**Comment by jbrooksuk** [comment]
I'm in the process of writing a test suite solely for the API so that we can be sure that this change does not break anything :)

**Comment by jbrooksuk** [comment]
Phew! This is why we have these new tests...

image

File: app/Http/Controllers/Api/ComponentController.php [hunk 1]
```diff
@@ -69,65 +56,77 @@ public function getComponent($id)
...
      */
-    public function putComponent($id)
+    public function putComponent(Component $component)
     {
-        $component = $this->component->update($id, Binput::except('tags'));
+        $component->update(Binput::except('tags'));
 
-        if (Binput::has('tags')) {
-            $tags = preg_split('/ ?, ?/', Binput::get('tags'));
+        if ($component->isValid('updating')) {
```
Comment (GrahamCampbell) [inline app/Http/Controllers/Api/ComponentController.php]: Better to throw the exception first, then avoid the nesting.

**Comment by joecohens** [comment]
@jbrooksuk Rebased and fixed and added more test to the Api. What else is needed to finish this PR?

**Comment by joecohens** [comment]
Lol fixed the tests locally don't know why travis if failing can it be configuration settings to run PHP server or something like that?

**Comment by jbrooksuk** [comment]
Lol, silly Travis.

**Comment by jbrooksuk** [comment]
Good to merge? All tests pass locally.

**Comment by GrahamCampbell** [comment]
It's because travis is rebasing before running the tests...

**Comment by GrahamCampbell** [comment]
Rebase this, and they will fail for you too, I think.

**Comment by jbrooksuk** [comment]
Strange behaviour?

**Comment by jbrooksuk** [comment]
Rebased, let's see what happens.

**Comment by jbrooksuk** [comment]
Why is Travis running twice anyway?

**Comment by jbrooksuk** [comment]
`EFFFEEEFEEFFFEEEFEFEEFEFFFEEEF` WTF, why?

**Comment by GrahamCampbell** [comment]
Probably someone's broken laravel.

**Comment by GrahamCampbell** [comment]
https://github.com/laravel/framework/blob/5.1/src/Illuminate/Support/helpers.php#L534

**Comment by GrahamCampbell** [comment]
https://github.com/laravel/framework/blob/5.1/src/Illuminate/Support/helpers.php#L538

**Comment by GrahamCampbell** [comment]
look, here, we never asign it back to itself

**Comment by GrahamCampbell** [comment]
Change that line to `$value = array_sort_recursive($value);` and see if it works now.

**Comment by GrahamCampbell** [comment]
> Why is Travis running twice anyway?

Because it runs once for the PR, and once for the push to this repo.

**Comment by GrahamCampbell** [comment]
If you want it to run only once, fork the repo to send PRs.

**Comment by jbrooksuk** [comment]
@GrahamCampbell the `$value` variable is by reference anyway, so it shouldn't need assigning back?

**Comment by GrahamCampbell** [comment]
I'm clutching at straws, lol.

**Comment by jbrooksuk** [comment]
This branch also closes #692 :)

**Pull request merged**